ui/pages/institutional/custody/custody.js

```diff
@@ -0,0 +1,645 @@
+    useState('');
+  const [selectedCustodianType, setSelectedCustodianType] = useState('');
+  const [connectError, setConnectError] = useState('');
+  const [currentJwt, setCurrentJwt] = useState('');
+  const [selectError, setSelectError] = useState('');
+  const [jwtList, setJwtList] = useState([]);
+  const [apiUrl, setApiUrl] = useState('');
+  const [addNewTokenClicked, setAddNewTokenClicked] = useState(false);
+  const [chainId, setChainId] = useState(0);
+  const [connectRequest, setConnectRequest] = useState(undefined);
+  const [accounts, setAccounts] = useState();
+  const [custodianButtons, setCustodianButtons] = useState(null);
+
+  useEffect(() => {
+    const custodianItems = [];
+    custodians.forEach((custodian) => {
+      if (
+        (!custodian.production &&
+          process.env.METAMASK_ENVIRONMENT === 'production') ||
+        custodian.hidden ||
+        (connectRequest &&
+          Object.keys(connectRequest).length &&
+          custodian.name !== selectedCustodianName)
+      ) {
+        return;
+      }
+
+      custodianItems.push(
+        <Box
+          key={uuidv4()}
+          display={DISPLAY.FLEX}
+          flexDirection={FLEX_DIRECTION.ROW}
+          justifyContent={JustifyContent.spaceBetween}
+          alignItems={AlignItems.center}
+          borderColor={BorderColor.borderDefault}
+          borderRadius={BorderRadius.SM}
+          padding={[3, 4]}
+          marginBottom={4}
+        >
+          <Box display={DISPLAY.FLEX} alignItems={AlignItems.center}>
+            {custodian.iconUrl && (
+              <img
+                width={32}
+                height={32}
+                src={custodian.iconUrl}
+                alt={custodian.displayName}
+              />
+            )}
+            <Text marginLeft={2}>{custodian.displayName}</Text>
+          </Box>
+
+          <Button
+            size={BUTTON_SIZES.SM}
+            data-testid="custody-connect-button"
+            onClick={async (_) => {
+              const jwtListValue = await dispatch(
+                mmiActions.getCustodianJWTList(custodian.name),
+              );
+              setSelectedCustodianName(custodian.name);
+              setSelectedCustodianType(custodian.type);
+              setSelectedCustodianImage(custodian.iconUrl);
+              setSelectedCustodianDisplayName(custodian.displayName);
+              setApiUrl(custodian.apiUrl);
+              setCurrentJwt(jwtListValue[0] || '');
+              setJwtList(jwtListValue);
+              trackEvent({
+                category: 'MMI',
+                event: 'Custodian Selected',
+                properties: {
+                  custodian: custodian.name,
+                },
+              });
+            }}
+          >
+            {t('select')}
+          </Button>
+        </Box>,
+      );
+    });
+
+    setCustodianButtons(custodianItems);
+  }, [connectRequest, custodians, dispatch, selectedCustodianName]);
```

Comment:
Why is this an effect that sets state instead of useMemo to memorize an array of items?

Reply:
In general setting a react tree in state is a bad practice.

Reply:
As I have said before, lets not set the precedent of breaking up long functions with render methods we already said we don't want to do that before we made the move to functional components.